### PR TITLE
Bloomchain

### PR DESCRIPTION
**Pull request by debris**
requires #996 and #997

small fixes, optimizations and removing duplicated logic

- `blockchain::BlockChainConfig` moved to its own file and renamed to just `Config`
- removed old `chainfilter` module and everything related to it, eg. `blockchain::BloomIndexer`. Use `bloomchain` crate instead.
- in extras database type byte is the first byte, instead being the last. That should improve query performance
- `BlocksBlooms` database keys are smaller - 6 bytes (33 bytes before).
- `BlockHashes` database keys are smaller - 5 bytes( 33 bytes before).
- change client database version to `6.0`

### CODE REVIEW

File: parity/rpc.rs [hunk 1]
```diff
@@ -113,6 +113,7 @@ fn setup_rpc_server(apis: Vec<&str>, deps: &Arc<Dependencies>) -> Server {
 				server.add_delegate(EthcoreClient::new(&deps.miner, deps.logger.clone(), deps.settings.clone()).to_delegate())
 			},
 			"traces" => {
+				// not adding to modules, since `traces` is not supported in geth
```
Comment (tomusdrw) [inline parity/rpc.rs]: Not needed.

Comment (debris) [inline parity/rpc.rs]: right! incorrect merge 👍

**Comment by gavofyork** [comment]
might be worth staging somewhere other than master or we'll be resyncing multiple times with the current DB changes...

**Comment by debris** [comment]
I can keep this pr up to date until there is a right moment to merge it

**Comment by NikVolf** [comment]
btw can't the database be auto-upgraded somehow?

**Comment by debris** [comment]
Added migrations

File: ethcore/src/blockchain/extras.rs [hunk 2]
```diff
@@ -30,97 +32,98 @@ pub enum ExtrasIndex {
...
-	fn index() -> ExtrasIndex;
-}
-
-impl ExtrasIndexable for H256 {
-	fn index() -> ExtrasIndex {
-		ExtrasIndex::BlockHash
+	let mut result = H264::default();
+	result[0] = i as u8;
+	unsafe {
+		ptr::copy(hash.as_ptr(), result.as_mut_ptr().offset(1), 32);
```
Comment (arkpar) [inline ethcore/src/blockchain/extras.rs]: prefer `clone_from_slice` to unsafe code.
```rust
let result_key = &mut result[1 .. ];
result_key.clone_from_slice(hash);
```

File: ethcore/src/migrations/extras/v6.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+pub struct ToV6;
+
+impl ToV6 {
+	fn migrate_old_key(&self, old_key: Vec<u8>, index: u8, len: usize) -> Vec<u8> {
+		let mut result = vec![];
+		result.reserve(len);
+		unsafe {
+			result.set_len(len);
+			result[0] = index;
+			ptr::copy(old_key.as_ptr().offset(33 - len as isize), result.as_mut_ptr().offset(1), len - 1);
```
Comment (arkpar) [inline ethcore/src/migrations/extras/v6.rs]: use `clone_from_slice`?

File: ethcore/src/trace/db.rs [hunk 4]
```diff
@@ -65,15 +76,15 @@ impl Deref for TraceGroupKey {
...
 
 	fn key(&self) -> Self::Target {
 		let mut result = [0u8; 6];
-		result[0] = TraceDBIndex::BlockTracesBloomGroups as u8;
-		result[1] = self.level;
+		result[0] = TraceDBIndex::BloomGroups as u8;
+		result[1] = self.0.level;
 		unsafe {
-			ptr::copy(&[self.index] as *const u32 as *const u8, result.as_mut_ptr().offset(2), 4);
+			ptr::copy(&[self.0.index] as *const u32 as *const u8, result.as_mut_ptr().offset(2), 4);
```
Comment (arkpar) [inline ethcore/src/trace/db.rs]: Another potential endianess problem? Pointer cast is evil

Comment (debris) [inline ethcore/src/trace/db.rs]: 👍 I forgot about it

File: ethcore/src/blockchain/blockchain.rs [hunk 5]
```diff
@@ -129,21 +108,35 @@ pub trait BlockProvider {
...
 }
 
 struct CacheManager {
 	cache_usage: VecDeque<HashSet<CacheID>>,
 	in_use: HashSet<CacheID>,
 }
 
+impl bc::group::BloomGroupDatabase for BlockChain {
+	fn blooms_at(&self, position: &bc::group::GroupPosition) -> Option<bc::group::BloomGroup> {
+
```
Comment (gavofyork) [inline ethcore/src/blockchain/blockchain.rs]: superfluous line

File: ethcore/src/blockchain/config.rs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Blockchain configuration.
+use bloomchain::Config as BloomConfig;
+
+/// Blockchain configuration.
+#[derive(Debug)]
+pub struct Config {
```
Comment (gavofyork) [inline ethcore/src/blockchain/config.rs]: is it really sensible to place runtime options (`_cache_size`) with database-persistent options (`log_blooms`)?

Comment (debris) [inline ethcore/src/blockchain/config.rs]: This option (`log_blooms`) can be configurable now. Syncing from scratch could be done with different number of `levels` and `elements_per_index`. There is no cli for it yet, but I'm willing to add it someday.

**Pull request merged**